cpp/test/sg/fil_test.cu

```diff
@@ -1069,4 +1072,104 @@ TEST_P(TreeliteThrowSparse8FilTest, Import) { check(); }
 INSTANTIATE_TEST_CASE_P(FilTests, TreeliteThrowSparse8FilTest,
                         testing::ValuesIn(import_throw_sparse8_inputs));
 
+template <typename T>
+__device__ void serial_multireduction(const T* in, T* out, int set_size,
+                                      int n_sets) {
+  __syncthreads();
+  if (threadIdx.x < set_size) {
+    int reduction_id = threadIdx.x;
+    T sum{};
+    for (int set = 0; set < n_sets; ++set)
+      sum += in[reduction_id + set * set_size];
+    out[reduction_id] = sum;
+  }
+  __syncthreads();
+}
+
+// the most threads a block can have
+const int max_threads = 1024;
+
+template <int R, typename T>
+__device__ void test_single_radix(const T value, const int valid_threads,
+                                  int* error) {
+  __shared__ T work[max_threads], correct_result[max_threads];
+  for (int set_size = 1; set_size < valid_threads; ++set_size) {
+    int n_sets = valid_threads / set_size;
+    work[threadIdx.x] = value;
+    serial_multireduction(work, correct_result, set_size, n_sets);
+    T sum = multireduction<R>(work, set_size, n_sets);
+    if (threadIdx.x < set_size && sum != correct_result[threadIdx.x]) {
+      printf("multireduction<%d>(on %d sets sized %d)[%d] gave wrong result\n",
+             R, n_sets, set_size, threadIdx.x);
+      atomicAdd(error, 1);
+    }
+  }
+}
+
+template <typename T>
+// TODO(levsnv): eventually replace args with std::spans
+__global__ void test_multireduction_k(T* data, int data_size, int* n_values,
+                                      int n_n_values, int* error) {
+  if (blockIdx.x >= n_n_values) return;
+  int my_size = n_values[blockIdx.x];
+  T value = threadIdx.x > my_size ? data[threadIdx.x] : T();
+  test_single_radix<2>(value, my_size, error);
```

Comment:
Wouldn't it be better to test different radix values in different thread blocks, to increase parallelism?